woeip/apps/air_quality/dustrak.py

```diff
@@ -0,0 +1,295 @@
+        raise NotImplementedError('Minute sampling intervals not supported')
+
+    sample_freq = sample_interval_seconds + 'S'
+    sample_times = pd.date_range(start_time, freq=sample_freq, periods=len(data), tz='UTC')
+
+    data['time'] = sample_times
+    data.sort_values(by='time', inplace=True)
+
+    return data
+
+
+def load_gps(path):
+    """Load and condition data from a GPS raw data file
+
+    Parameters
+    ----------
+    path : str
+
+    Returns
+    -------
+    A pandas DataFrame
+    """
+    with open(path, 'r') as f:
+        gps = []
+        for sample in f:
+            if sample.startswith('$GPRMC'):
+                gps_sample = parse_gps_sentence(sample)
+                gps_dict = sentence_to_dict(gps_sample)
+                gps.append(gps_dict)
+
+    gps = pd.DataFrame(gps)
+
+    sample_times = combine_date_and_time(gps.datestamp, gps.timestamp)
+    sample_times = pd.DatetimeIndex(sample_times, tz='UTC')
+    gps['time'] = sample_times
+    gps.sort_values(by='time', inplace=True)
+
+    gps['lat'] = gps.lat.apply(float)
+    gps['lon'] = gps.lon.apply(float)
+
+    gps['lat'] = gps.lat.apply(degree_minute_to_decimal)
+    gps['lon'] = gps.lon.apply(degree_minute_to_decimal)
+
+    latitudes = []
+    for _, (lat, lat_dir) in gps[['lat', 'lat_dir']].iterrows():
+        if lat_dir == 'N':
+            latitudes.append(lat)
+        else:
+            latitudes.append(-lat)
+
+    longitudes = []
+    for _, (lon, lon_dir) in gps[['lon', 'lon_dir']].iterrows():
+        if lon_dir == 'E':
+            longitudes.append(lon)
+        else:
+            longitudes.append(-lon)
+
+    gps['lat'] = latitudes
+    gps['lon'] = longitudes
+
+    return gps
+
+
+def join(dustrak_path, gps_path, tz='America/Los_Angeles', tolerance=3.):
+    """Join a DusTrak file and a GPS file into a single table.
+
+    The DusTrak device collects time stamps and air quality measurements, but no geospatial
+    information. Scientists will also take a GPS device on their sessions that records time stamps
+    and latitudes/longitudes. Those two data files need to be combined.
+
+    Parameters
+    ----------
+    dustrak_path : str
+    gps_path : str
+    tz : str
+    tolerance : numeric
+        How far away (in seconds) can a air quality measurement be from a GPS measurement to be
+        linked to it
+
+    Returns
+    -------
+    A pandas DataFrame containing the sample time (in UTC), latitude, longitude, and measurement
+    """
+    data = load_dustrak(dustrak_path, tz)
+    gps = load_gps(gps_path)
+
+    joined_data = pd.merge_asof(data, gps, on='time', direction='nearest',
+                                tolerance=pd.Timedelta(f'{tolerance}s'))
+
+    invalid_indices = joined_data[['lat', 'lon', 'Mass [mg/m3]']].isnull().any(1)
+    joined_data = joined_data[~invalid_indices]
+
+    n_dropped = invalid_indices.sum()
+
+    if n_dropped > 0:
+        message = (f"""{n_dropped} air quality samples dropped that were """
```

Comment:
This definitely fits on a single line: `message = f"{n_dropped} air quality samples dropped that were not within {tolerance} seconds of a GPS sample."`.